src/smt/optimization_solver.h

```diff
@@ -0,0 +1,118 @@
+/*********************                                                        */
+/*! \file optimization_solver.h
+ ** \verbatim
+ ** Top contributors (to current version):
+ **   Michael Chang
+ ** This file is part of the CVC4 project.
+ ** Copyright (c) 2009-2020 by the authors listed in the file AUTHORS
+ ** in the top-level source directory) and their institutional affiliations.
+ ** All rights reserved.  See the file COPYING in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** \brief The solver for optimization queries
+ **/
+
+#include "cvc4_private.h"
+
+#ifndef CVC4__SMT__OPTIMIZATION_SOLVER_H
+#define CVC4__SMT__OPTIMIZATION_SOLVER_H
+
+#include "expr/node.h"
+#include "expr/type_node.h"
+#include "smt/assertions.h"
+#include "util/result.h"
+
+namespace CVC4 {
+namespace smt {
+
+/**
+ * An enum for optimization queries.
+ *
+ * Represents whether an objective should be minimized or maximized
+ */
+enum CVC4_PUBLIC ObjectiveType
+{
+  OBJECTIVE_MINIMIZE,
+  OBJECTIVE_MAXIMIZE,
+  OBJECTIVE_UNDEFINED
+};
+
+/**
+ * An enum for optimization queries.
+ *
+ * Represents the result of a checkopt query
+ * (unimplemented) OPT_OPTIMAL: if value was found
+ */
+enum CVC4_PUBLIC OptResult
+{
+  // the original set of assertions has result UNKNOWN
+  OPT_UNKNOWN,
+  // the original set of assertions has result UNSAT
+  OPT_UNSAT,
+  // the optimization loop finished and optimal
+  OPT_OPTIMAL,
+
+  // The last two values are here as a preparation for future work
+  // in which partial and approximate optimizations will be supported.
+
+  // if the optimized value is only partial
+  OPT_SAT_PARTIAL,
+  // if the solver halted early and value is only approximate
+  OPT_SAT_APPROX
```

Comment:
what would be the difference between `OPT_SAT_PARTIAL` and `OPT_SAT_APPROX`?

Reply:
As far as I understand, OPT_SAT_PARTIAL is for when the solver times out and the answer is not well-bounded
whereas OPT_SAT_APPROX is for when the answer is well-bounded but either not discrete or not converging for some other reason and we return an approximate value.

Reply:
hmmm, if the solver times out without a single sat result in the optimization loop then the result shouldn't be OPT_SAT_PARTIAL, instead, it should be OPT_UNKNOWN. if it does find at least one sat result in the optimization loop then the result is OPT_SAT_APPROX. So, I don't see a use case for OPT_SAT_PARTIAL.

Reply:
Hmm. I see where you're coming from. 
OPT_UNKOWN is reserved for cases where the underlying sat solver times out IE when OPT_UNKNOWN is returned from the solver in the optimization loop. 
OPT_SAT_PARTIAL is returned if we were to get stuck in our optimization loop for some reason, and want to end it prematurely. The underlying solver could return only SAT, but our solver may never converge in which case we would want to return OPT_SAT_PARTIAL.
please lmk if you want more info and we can set up a time to meet.

Reply:
In the last case which you mentioned, it should be ok to return OPT_SAT_APPROX.